src/net/server/handlers/create_query.rs

```diff
@@ -0,0 +1,246 @@
+#[pin_project]
+pub struct ByteArrStream {
+    #[pin]
+    body: BodyStream,
+    size_in_bytes: u32,
+    buffered_size: u32,
+    buffered: Vec<Bytes>,
+}
+
+impl ByteArrStream {
+    pub fn new(body: BodyStream, size_in_bytes: u32) -> Self {
+        Self {
+            body,
+            size_in_bytes,
+            buffered_size: 0,
+            buffered: Vec::new(),
+        }
+    }
+
+    /// returns [`Bytes`] of `size_in_bytes` length from the buffered chunks
+    /// # Panics
+    /// if the total length of byte buffers is smaller than `size_in_bytes`
+    fn take_bytes(size_in_bytes: u32, buffered: &mut Vec<Bytes>, buffered_size: &mut u32) -> Bytes {
+        assert!(*buffered_size >= size_in_bytes);
+
+        // take from single buffer
+        // simply splits the first buffer in-place
+        if buffered[0].len() > size_in_bytes as usize {
+            let out_bytes = buffered[0].split_to(size_in_bytes as usize);
+            *buffered_size -= size_in_bytes;
+            out_bytes
+        } else {
+            // take across buffers
+            // removes expended buffers after taking
+            let mut out_bytes = Vec::new();
+            let mut remaining_bytes = size_in_bytes as usize;
+            let mut i = 0;
+            loop {
+                if remaining_bytes == 0 {
+                    break;
+                }
+                let buffer = &mut buffered[i];
+                let buffer_len = buffer.len();
+                // current `buffer` has more bytes than needed
+                if buffer_len > remaining_bytes {
+                    out_bytes.push(buffer.split_to(remaining_bytes));
+                    remaining_bytes = 0;
+                } else {
+                    // current `buffer` has <= bytes needed, needs removal
+                    out_bytes.push(buffer.clone());
+                    i += 1;
+                    remaining_bytes -= buffer.len();
+                }
+            }
+
+            // remove expended buffers
+            buffered.drain(0..i);
+            // reduce size of total buffers accordingly
+            *buffered_size -= size_in_bytes;
+            // create new `Bytes` that combines chunks
+            let bytes_vec = out_bytes.into_iter().fold(
+                Vec::with_capacity(size_in_bytes as usize),
+                |mut acc, n| {
+                    acc.extend_from_slice(&n);
+                    acc
+                },
+            );
+            Bytes::from(bytes_vec)
+        }
+    }
+}
+
+impl Stream for ByteArrStream {
+    type Item = Result<Bytes, MpcHelperServerError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.as_mut().project();
+        loop {
+            // if we currently have enough bytes, return them
+            if *this.buffered_size >= *this.size_in_bytes {
+                return Poll::Ready(Some(Ok(Self::take_bytes(
```

Comment:
this pattern is weird, why don't you take `&mut self` and access/update state there? You pinned yourself already

Reply:
I'm not sure how to do this because the first line in `poll_next` is `let mut this = self.as_mut().project();`, and I don't know how to pass that type.

Reply:
https://github.com/taiki-e/pin-project-lite/issues/10

Reply:
took a stab at it, works. I would also suggest to change `take_bytes` contract to return `Option<Bytes>` - in this case your stream implementation wouldn't be exposed to non-relevant details. Only thing it needs to know is whether struct is ready to send more bytes or no. 

```rust
        let mut this = self.as_mut().project();
        loop {
            // if we currently have enough bytes, return them
            if *this.buffered_size >= *this.size_in_bytes {
                return Poll::Ready(Some(Ok(self.take())));
            }
```

Reply:
👍 makes sense

Reply:
You don't need `this` imo if your buffer is ready, so you can project only when you need to poll pinned stream, after this line

Reply:
yep, that's what i ended up doing